src/str/similar.rs

```diff
@@ -0,0 +1,265 @@
+    limit: SimilarLimit,
+}
+
+impl SimilarPredicate {
+    /// The maximum allowed number of changes.
+    ///
+    /// Default: `0`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use predicates::prelude::*;
+    /// use predicates::str::similar3;
+    ///
+    /// let s1 = "Hello World!";
+    /// let s2 = "Hello... World?";
+    ///
+    /// assert_eq!(false, similar3(s1).eval(s2));
+    /// assert_eq!(true, similar3(s1).changes(2).eval(s2));
+    /// assert_eq!("changes(2)", &similar3(s1).changes(2).find_case(true, s2).unwrap().product_value("measure").unwrap());
+    /// ```
+    pub fn changes(mut self, changes: u32) -> Self {
+        self.limit = SimilarLimit::Changes(changes);
+        self
+    }
+
+    /// The minimum required [similarity](similar::get_diff_ratio).
+    ///
+    /// Default: `1.0`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use predicates::prelude::*;
+    /// use predicates::str::similar3;
+    ///
+    /// let s1 = "Hello World!";
+    /// let s2 = "Hello... World?";
+    ///
+    /// assert_eq!(false, similar3(s1).eval(s2));
+    /// assert_eq!(true, similar3(s1).ratio(0.8).eval(s2));
+    /// assert_eq!("ratio(0.8148148)", &similar3(s1).find_case(false, s2).unwrap().product_value("measure").unwrap());
+    /// ```
+    pub fn ratio(mut self, ratio: f32) -> Self {
+        // FIXME: clamp to 0..1 ? Panic ?
+        self.limit = SimilarLimit::Ratio(ratio);
+        self
+    }
+
+    /// The [diffing algorithm](similar::Algorithm).
+    ///
+    /// Default: [`Myers`](similar::Algorithm::Myers)
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use predicates::prelude::*;
+    /// use predicates::str::similar3;
+    /// use similar::Algorithm::*;
+    ///
+    /// let s1 = "Hello World!";
+    /// let s2 = "Hello... World?";
+    /// for (pred, expect) in vec![(similar3(s1), "ratio(0.8148148)"),
+    ///                            (similar3(s1).algorithm(Myers), "ratio(0.8148148)"),
+    ///                            (similar3(s1).algorithm(Lcs), "ratio(0.44444445)"),
+    ///                            (similar3(s1).algorithm(Patience), "ratio(0.8148148)")] {
+    ///   let m = pred.find_case(false, s2).unwrap().product_value("measure").unwrap();
+    ///   assert_eq!(expect, &m);
+    /// }
+    /// ```
+    pub fn algorithm(mut self, alg: similar::Algorithm) -> Self {
+        self.algorithm = alg;
+        self
+    }
+
+    fn new(old: borrow::Cow<'static, str>, op: SimilarOp) -> Self {
+        Self {
+            old,
+            algorithm: similar::Algorithm::Myers,
+            op,
+            limit: SimilarLimit::Ratio(1.0),
+        }
+    }
+
+    fn run(&self, new: &str) -> (bool, Vec<similar::DiffOp>, SimilarLimit) {
+        use similar::DiffOp::*;
+        use SimilarLimit::*;
+
+        let chunks =
+            similar::capture_diff_slices(self.algorithm, self.old.as_bytes(), new.as_bytes());
+        match self.limit {
+            Ratio(r) => {
```

Comment:
Any thoughts on usefulness of `ratio`? Particularly, since I'm unsure of `changes` being in

Reply:
Hum, if we have neither `ratio` nor `changes` then this predicate becomes the same as a basic string equality predicate except with a nice diff output ? If that's all we want, we should merge the diff functionality into the basic equality predicate. It can still be feature-gated, just compute and format the diff only if strings are not equal.

Reply:
Well, still need to discuss `split` :)

There is a part of me that would want them distinct because I feel like they are different use cases.  A one word string probably wouldn't want to use `similar`, the output would just get too noisy.  Granted, I can see having a helper to turn a string predicate into a diff predicate.